### PR TITLE
OLAS Connector [wip]

### PR DESCRIPTION
**Pull request by blakewatters**
Temporary PR to have a look at the connector state.

### CODE REVIEW

File: servo/connectors/olas/controller.py [hunk 1]
```diff
@@ -0,0 +1,880 @@
...
+            ctl = importlib.reload(module)
+            self.__class__ = ctl.OLASController
+            self.mode_table[mode.CPU_PREDICT] = getattr(self, "scale_by_cpu_predict")
+            servo.logger.info(f"{module} new classid: {id(self.__class__)}")
+
+            importlib.reload(prometheus)
+            importlib.reload(envoy_stats)
+            importlib.reload(client)
+            importlib.reload(kube)
+        self.reload_mtime = mtime
```
Comment (blakewatters) [inline servo/connectors/olas/controller.py:880]: This is a bad implementation of code reloading. I would recommend [Watchgod](https://pypi.org/project/watchgod/) instead

File: servo/connectors/olas/envoy_stats.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import asyncio
+
+import aiohttp
+from prometheus_client.parser import text_string_to_metric_families
+
+import servo
+
+
+class Histogram:
```
Comment (blakewatters) [inline servo/connectors/olas/envoy_stats.py:9]: This is on me to reimplement within the ServoX prometheus client

File: servo/connectors/olas/envoy_stats.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+class EnvoyStats:
+    def __init__(self, body):
+        metrics = {}
+        ms = text_string_to_metric_families(body)
+        for family in ms:
+            for sample in family.samples:
+                # if sample.name.endswith('upstream_rq_time_bucket'):
+                #    print (family.name, sample.name, sample.labels, sample.value)
+                metrics.setdefault(sample.name, []).append((sample.labels, sample.value))
```
Comment (blakewatters) [inline servo/connectors/olas/envoy_stats.py:65]: This seems like this is a missing sub-model

File: servo/connectors/olas/envoy_stats.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    def get_delta_bucket(self, name, label={}, prev=[]):
+        total = self.get_buckets(name, label)
+        # servo.logger.info(f'name {name} label {label}\ntotal_bucket {pprint.pformat(total)}\nprev {pprint.pformat(prev)}')
+        if not prev:
+            return total, total
+        return self.delta_bucket(prev, total), total
+
+    def get_histogram(self, name, label={}, prev=[]):
+        delta, total = self.get_delta_bucket(name, label, prev)
+        return Histogram(delta), total
```
Comment (blakewatters) [inline servo/connectors/olas/envoy_stats.py:82]: These arguments need to be explained. I can't follow what is going on here

File: servo/connectors/olas/envoy_stats.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    def delta_bucket(self, old, new):
+        assert len(old) == len(new)
+        delta = [(n[0], n[1] - o[1]) for o, n in zip(old, new)]
+        return delta
+
+    def merge_bucket(self, ba, bb):
+        assert len(ba) == len(bb)
+        return [(a[0], a[1] + b[1]) for a, b in zip(ba, bb)]
```
Comment (blakewatters) [inline servo/connectors/olas/envoy_stats.py:91]: Seems like there should be a `Bucket` class that supports addition and subtraction methods

File: servo/connectors/olas/envoy_stats.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    def get_sample(self, name, label):
```
Comment (blakewatters) [inline servo/connectors/olas/envoy_stats.py:93]: None is a valid return value?

File: servo/connectors/olas/envoy_stats.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        return (v - prev) / interval if interval else 0, v
+
+
+async def get_envoy_stats(url, session):
+    '''return EnvoyStats object'''
+    async with session.get(url, timeout=1) as res:
+        if res.status == 200:
+            return EnvoyStats(res.text), ''
+    servo.logger.error("Error envoy stats fetch error %s" % (url))
+    return None, "Fetch failed"
```
Comment (blakewatters) [inline servo/connectors/olas/envoy_stats.py:111]: This should be raising an exception. Returning a tuple of `(value || None), (Error Message || '')` does not make sense.

File: servo/connectors/olas/envoy_stats.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        v = self.get_sample(name, label)
+        return (v - prev) / interval if interval else 0, v
+
+
+async def get_envoy_stats(url, session):
+    '''return EnvoyStats object'''
+    async with session.get(url, timeout=1) as res:
+        if res.status == 200:
+            return EnvoyStats(res.text), ''
+    servo.logger.error("Error envoy stats fetch error %s" % (url))
```
Comment (blakewatters) [inline servo/connectors/olas/envoy_stats.py:110]: Use an f-string

File: servo/connectors/olas/envoy_stats.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    return None, "Fetch failed"
+
+
+async def get_envoy_stats_nodes(baseurls):
+    '''
+    return a list of metrics from list of urls.
+    '''
```
Comment (blakewatters) [inline servo/connectors/olas/envoy_stats.py:117]: This comment has zero value.

File: servo/connectors/olas/eventqueue.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+        for queue in self.queues.get(event_type, []):
+            queue.put_nowait((event_type, event))
+
+    def subscribe_event(self, event_type):
+        queue = asyncio.Queue()
+        self.queues.setdefault(event_type, []).append(queue)
+        return queue
+
+
+eventQueue = EventQueue()
```
Comment (blakewatters) [inline servo/connectors/olas/eventqueue.py:19]: This whole class should be dropped.

File: servo/connectors/olas/fastpath.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+
+    def __init__(self, data, foresight=1, window_size=5):
+        self.data_lake = self.build_data_lake(data)
+        self.window_size = int(window_size)
+        self.foresight = int(foresight)
+        if self.foresight < 1:
+            raise ValueError(
+                "Fastpath Argument Error: "
+                "Lowest resolution supported for prediction is 1 min. "
+                "Please modidy argument for foresight.")
```
Comment (blakewatters) [inline servo/connectors/olas/fastpath.py:29]: So foresight is expressed in terms of minutes but everything else is in seconds?

File: servo/connectors/olas/fastpath.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        - foresight: in mins, denotes how much in advance the prediction will be,
+                     and also what frequency used to downsample raw data.
+                     Should be larger than the cadence of received measurements.
+        - window_size: window size used to smooth resampled time series
+    '''
+
+    def __init__(self, data, foresight=1, window_size=5):
+        self.data_lake = self.build_data_lake(data)
+        self.window_size = int(window_size)
+        self.foresight = int(foresight)
```
Comment (blakewatters) [inline servo/connectors/olas/fastpath.py:24]: What values are we anticipating can be passed in that are not integers but can be coerced into them?

File: servo/connectors/olas/fastpath.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        self.resample()
+        # Window_size indicats half width of intended sliding window,
+        # thus 4 in below condition is chosen to guarantee at least
+        # half of resampled data could be applied with hampel_filter.
+        if len(self.traffic) < 4 * self.window_size:
+            raise ValueError(
+                "Fastpath Argument Error: "
+                "Initial data provided are not commensurate with window size. "
+                "Please either decrease window size or increase data quantity. "
+                "Suggested time span for initial data should be at least 4*window_size*foresight mins.")
```
Comment (blakewatters) [inline servo/connectors/olas/fastpath.py:40]: This constraint says "Suggest" but its actually enforced as a validation. Needs to be documented and explained.

File: servo/connectors/olas/fastpath.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+    def build_data_lake(self, data):
```
Comment (blakewatters) [inline servo/connectors/olas/fastpath.py:42]: How is this a data lake?

File: servo/connectors/olas/kube.py [hunk 15]
```diff
@@ -0,0 +1,543 @@
+import json
+import re
+import time
+from typing import List
+
```
Comment (blakewatters) [inline servo/connectors/olas/kube.py:5]: I will drill in here and implement any missing functionality in the ServoX Kubernetes core. Assume all of this just goes away

File: servo/connectors/olas/prometheus.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+    data: PrometheusData
+    status: str
+
+
+class PrometheusVector(BaseModel):
+    metric: Dict
+    value: List
+
+
+class Histogram:
```
Comment (blakewatters) [inline servo/connectors/olas/prometheus.py:27]: I will implement Histogram support in the ServoX core

File: servo/connectors/olas/server_classes.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+class Message(BaseModel):
+    ts: float
+    msg: str
+
+
+class Prediction(BaseModel):
+    src: str
+
+
+class TimeSeries(BaseModel):
```
Comment (blakewatters) [inline servo/connectors/olas/server_classes.py:30]: How is this a time series? It looks like a single data point.

File: servo/connectors/olas/server_classes.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+class PodMetrics(BaseModel):
+    name: str
+    allocation: int = 0
+    node: int = 0
+    metrics: Dict[str, float]
+
+
+class Metrics(BaseModel):
+    ts: float
```
Comment (blakewatters) [inline servo/connectors/olas/server_classes.py:89]: Probably should be a `datetime.datetime`

File: servo/connectors/olas/server_classes.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    excluded_pods: List[PodMetrics] = []
+    deployments: List[str] = []
+    excluded_deployments: List[str] = []
+    replicas: int = 0
+
+
+class GetMetrics(BaseModel):
+    start_ts: float
+    stop_ts: float = 0
+    limit: int = 8192
```
Comment (blakewatters) [inline servo/connectors/olas/server_classes.py:104]: What is this number about?

File: servo/connectors/olas/server_classes.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+    pods: List[PodMetrics] = []
+    excluded_pods: List[PodMetrics] = []
+    deployments: List[str] = []
+    excluded_deployments: List[str] = []
+    replicas: int = 0
+
+
+class GetMetrics(BaseModel):
+    start_ts: float
+    stop_ts: float = 0
```
Comment (blakewatters) [inline servo/connectors/olas/server_classes.py:103]: Should these be `datetime.datetime` objects? Why does stop have a default of zero? Should it optional?

File: servo/connectors/olas/server_classes.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    allocations: List[Allocation] = []
+    nodes: List[NodeMetrics] = []
+    pods: List[PodMetrics] = []
+    excluded_pods: List[PodMetrics] = []
+    deployments: List[str] = []
+    excluded_deployments: List[str] = []
+    replicas: int = 0
+
+
+class GetMetrics(BaseModel):
```
Comment (blakewatters) [inline servo/connectors/olas/server_classes.py:101]: Why is this model named with a verb prefix?

File: servo/connectors/olas/server_classes.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+
+class NodeType(BaseModel):
+    name: str
+    cpu_allocatable: float
+    memory_allocatable: float
+    cpu_capacity: float
+    memory_capacity: float
+
+
+class Usage(BaseModel):
```
Comment (blakewatters) [inline servo/connectors/olas/server_classes.py:69]: I'd probably call this `Utilization`

File: servo/connectors/olas/server_classes.py [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+
+
+class Allocation(BaseModel):
+    cpu_request: float = 0
+    cpu_limit: float = 0
+    memory_request: float = 0
+    memory_limit: float = 0
+
+
+class NodeType(BaseModel):
```
Comment (blakewatters) [inline servo/connectors/olas/server_classes.py:61]: How is this a node type?

File: servo/connectors/olas/server_classes.py [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+    name: str
+    cpu_allocatable: float
+    memory_allocatable: float
+    cpu_capacity: float
+    memory_capacity: float
```
Comment (blakewatters) [inline servo/connectors/olas/server_classes.py:66]: Seems like there's a missing sub-model that could encapsulate these values.

Is this supposed to be a range?

File: servo/connectors/olas/server_classes.py [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+class TrainPodModel(BaseModel):
+    start: float
+    end: float
+
+
+class Allocation(BaseModel):
+    cpu_request: float = 0
+    cpu_limit: float = 0
+    memory_request: float = 0
+    memory_limit: float = 0
```
Comment (blakewatters) [inline servo/connectors/olas/server_classes.py:58]: `servo.connectors.kubernetes.CPU` and `servo.connectors.kubernetes.Memory` should cover this I think

File: servo/connectors/olas/prometheus.py [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+                if not result:
+                    return 0
+                metric = PrometheusVector.parse_obj(reply.data.result[0])
+                return float(metric.value[-1])
+            elif resultType == 'matrix':
+                # More than one vector value. Leave to the caller to process it.
+                return result
+            else:
+                servo.logger.error(f'Unhandled promtheus result type "{resultType}"')
+        except Exception:
```
Comment (blakewatters) [inline servo/connectors/olas/prometheus.py:115]: We can't be doing blanket exception handlers like this. If you don't know what could go wrong you cannot just suppress it, return an arbitrary value, and hope for this best.

**Review by blakewatters — COMMENTED** [review]
An extensive line-by-line close reading of the current reality of OLAS.

I haven't prioritized anything and I know that there is a lot here but I just went top to bottom and this is where we are at.

File: servo/connectors/olas/configuration.py [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+    def is_custom_scale_target(cls, values: Dict[str, str]) -> Dict[str, str]:
+        kind = values.get('kind')
+        if not kind or kind == 'Deployment':
+            return values
+
+        api = values.get('apiVersion')
+        if not api:
+            raise ValueError(f"ScaleTargetRef.apiVersion not specified for kind {kind}")
+        if len(api.split('/')) != 2:
+            raise ValueError(f"ScaleTargetRef.apiVersion {api} should contain exactly one '/'")
```
[the remark this answers is not in the document]
Comment (pnickolov) [reply servo/connectors/olas/configuration.py:31]: @blakewatters the objective here is to mimic the k8s convention used in HPA, e.g., https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/ search for scaleTargetRef.

File: servo/connectors/olas/connector.py [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+    license=servo.License.apache2,
+    maturity=servo.Maturity.experimental,
+)
+class OLASConnector(servo.BaseConnector):
+    config: OLASConfiguration
+
+    @servo.on_event()
+    def startup(self) -> None:
+        loop = asyncio.get_event_loop()
+        asyncio.set_event_loop(loop)
```
[the remark this answers is not in the document]
Comment (pnickolov) [reply servo/connectors/olas/connector.py:26]: @theStrenuous this should be fixed in order to fit properly in ServoX

File: servo/connectors/olas/connector.py [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+
+        olas = controller.OLASController(cfg_dict, backend_dict)
+        try:
+            task = loop.create_task(_olas_task(olas))
+            loop.run_until_complete(task)
+        except Exception:
+            servo.logger.critical(traceback.format_exc())
+            task = loop.create_task(olas.client.upload_message(olas.ts, traceback.format_exc()))
+            loop.run_until_complete(task)
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (pnickolov) [reply servo/connectors/olas/connector.py:45]: @theStrenuous this should be fixed to fit properly and work; please restore the repeating task functionality as the regular event source. Rely on ServoX's main to operate the event loop

File: servo/connectors/olas/controller.py [hunk 30]
```diff
@@ -0,0 +1,880 @@
...
+        with self.db.iterator(reverse=True, include_value=False) as it:
+            for k in it:
+                servo.logger.info(f"last key is {int(k)}")
+                return int(k)
+        return 0
+
+    def find_name_in_values(self, name, values):
+        for value in values:
+            if value.name == name:
+                return value.value
```
[the remark this answers is not in the document]
Comment (pnickolov) [reply servo/connectors/olas/controller.py:250]: Arguably, the former may be more readable. A more Pythonic implementation is likely possibly without going functional. What would be gained?

File: servo/connectors/olas/controller.py [hunk 31]
```diff
@@ -0,0 +1,880 @@
...
+                self.prometheus = PrometheusPlugin(plugin)
+            elif plugin.plugin == 'contour':
+                return
+                self.contour = ContourPlugin(plugin)
+
+    async def get_kube_node_metrics(self, pods):
+        count = 0
+        nodes = set(p.spec.node_name for p in pods)
+        nodes = [n for n in nodes if n]
+        while 1:
```
[the remark this answers is not in the document]
Comment (pnickolov) [reply servo/connectors/olas/controller.py:264]: @theStrenuous ^ ?

File: servo/connectors/olas/controller.py [hunk 32]
```diff
@@ -0,0 +1,880 @@
...
+            return
+
+        if not hpa:
+            return
+        name = hpa.spec.scale_target_ref.name
+        if not name.endswith("-disabled-by-olas"):
+            servo.logger.info(f"Found HPA {hpa.spec}")
+            hpa.spec.scale_target_ref.name = name + "-disabled-by-olas"
+            servo.logger.info(f"Disabling HPA {hpa.spec.scale_target_ref}")
+            await kube.patch_autoscaler(self.deployment, self.cfg.scaleTargetRef.namespace, hpa)
```
[the remark this answers is not in the document]
Comment (pnickolov) [reply servo/connectors/olas/controller.py:753]: Feature was requested by customer, as their automation generates the HPA automatically. This is opt-in

File: servo/connectors/olas/controller.py [hunk 33]
```diff
@@ -0,0 +1,880 @@
...
+            replicas, raw, reason = self.scale_request[-1]
+            await self.output_scale_with_constrain(replicas, raw)
+        self.last_control_ts = self.ts
+
+    async def process_event_task(self):
+        queue = self.eventqueue
+        while 1:
+            event_type, metrics = await queue.get()
+            servo.logger.info("Process event {metrics}")
+            await self.control_with_metrics(metrics)
```
[the remark this answers is not in the document]
Comment (pnickolov) [reply servo/connectors/olas/controller.py:837]: @theStrenuous this should be fixed

File: servo/connectors/olas/controller.py [hunk 34]
```diff
@@ -0,0 +1,880 @@
...
+
+    async def control_task(self):
+
+        self.boot()
+        self.eventqueue = eventqueue.eventQueue.subscribe_event("OLASMetrics")
+        asyncio.create_task(self.process_event_task())
```
[the remark this answers is not in the document]
Comment (pnickolov) [reply servo/connectors/olas/controller.py:843]: @theStrenuous please consider

File: servo/connectors/olas/controller.py [hunk 35]
```diff
@@ -0,0 +1,880 @@
...
+        pods = await self.get_pods()
+        await self.collect_metrics(pods)
+
+        while 1:
+            await self.sleep_loop_interval()
+            if self.cfg.config.devel:
+                self.try_reload()
+            pods = await self.get_pods()
+            metrics = await self.collect_metrics(pods)
+            eventqueue.eventQueue.put_event("OLASMetrics", metrics)
```
[the remark this answers is not in the document]
Comment (pnickolov) [reply servo/connectors/olas/controller.py:861]: @theStrenuous needs to be fixed. We can discuss with @blakewatters if the repeating task does not achieve what we need (I recall the timing here may be driven by arrival of metrics, not time)

File: servo/connectors/olas/controller.py (hunk 1 above)
Comment (pnickolov) [reply servo/connectors/olas/controller.py:880]: @theStrenuous I recommend disabling this until / unless ServoX has a mechanism for this

File: servo/connectors/olas/eventqueue.py (hunk 10 above)
Comment (pnickolov) [reply servo/connectors/olas/eventqueue.py:19]: @theStrenuous please fix as part of cleanup of the event loop

File: servo/connectors/olas/fastpath.py (hunk 12 above)
Comment (pnickolov) [reply servo/connectors/olas/fastpath.py:24]: String, float

**Pull request closed without merging**